### PR TITLE
Create a SONiC configuration management service

### PR DESCRIPTION
**Pull request by rajendra-dendukuri**
- Configuration initialization and migration service for SONiC

Refer to https://github.com/Azure/SONiC/pull/433 for more information.
 
Signed-off-by: Rajendra Dendukuri <rajendra.dendukuri@broadcom.com>

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by rajendra-dendukuri** [comment]
retest broadcom please

File: files/image_config/config-setup/config-setup [hunk 1]
```diff
@@ -0,0 +1,370 @@
+#!/bin/bash
+###########################################################################
+#Copyright 2019 Broadcom. All rights reserved. The term "Broadcom" refers #
+#to Broadcom Inc. and/or its subsidiaries.                                #
```
Comment (lguohan) [inline files/image_config/config-setup/config-setup]: should be apache license here.

File: files/image_config/config-setup/config-setup [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+      fi
+  fi
+
+  rm -f /tmp/pending_config_migration
+}
+
+# Take a backup of current SONiC configuration
+do_config_backup()
+{
+  echo "Taking backup of curent configuration"
```
Comment (lguohan) [inline files/image_config/config-setup/config-setup]: can you make tabstop=4?

Comment (rajendra-dendukuri) [reply files/image_config/config-setup/config-setup]: Will do. Waiting for more comments, before I make this cosmetic change.

File: files/image_config/config-setup/config-setup [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    fi
+
+    return $exit_status
+}
+
+# Reload minigraph.xml file on disk
+reload_minigraph()
+{
+    echo "Reloading minigraph..."
+    if [ ! -f /etc/sonic/init_cfg.json ]; then
```
Comment (rameshsanth) [inline files/image_config/config-setup/config-setup]: can't the command "config reload or config load -m <>" be used ?
This code here replicated steps from load_minigraph() function in sonic-utilities/config/main.py
If someone updates that part, they should also be updating it here. Doing it like this opens doorway to config load getting out of sync between "config load" and through ztp

Comment (rajendra-dendukuri) [reply files/image_config/config-setup/config-setup]: @rameshsanth

The scope of this change was to move the code from graphservice to config-setup service keeping the functionality as is for most part. Changing logic to use "config load" and  "config load_mingraph" can be done as part of a different PR.

**Comment by lguohan** [comment]
retest vsimage please

File: files/image_config/config-setup/config-setup (hunk 3 above)
Comment (lguohan) [reply files/image_config/config-setup/config-setup]: I agree with @rameshsanth , created an issue to merge load_minigraph in config-setup and update_minigraph and assigned to you. @rajendra-dendukuri

**Comment by keboliu** [comment]
@rajendra-dendukuri any suggestion about how to add a customer hook scripts? by the way, I didn't see a folder "/etc/config-setup/" and it's subfolders on a sonic switch with the latest 201911 image installed?

**Comment by rajendra-dendukuri** [comment]
> @rajendra-dendukuri any suggestion about how to add a customer hook scripts? by the way, I didn't see a folder "/etc/config-setup/" and it's subfolders on a sonic switch with the latest 201911 image installed?

@keboliu That is correct. The provision has been added but there are no hook scripts that have been added as of now. Something like below can be added to files/build_templates/sonic_debian_extension.j2. https://gist.github.com/rajendra-dendukuri/8481dc4025b8329416cfd91deca2d1a6

File: files/image_config/config-setup/config-setup [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+    return 0
+}
+
+# Create SONiC configuration for first time bootup
+#  - If ZTP is enabled, ZTP configuraion is created
+#  - If ZTP is disabled and updategraph is disabled, factory default configuration
+#    is created
+#  - If updategraph is enabled and ZTP is disabled, updategraph initializes
+#    configuration
+do_config_intialization()
```
Comment (wendani) [inline files/image_config/config-setup/config-setup:251]: intialization looks like a typo

Comment (rajendra-dendukuri) [reply files/image_config/config-setup/config-setup:251]: Thanks for pointing it out. I created below PR to fix this issue.

https://github.com/Azure/sonic-buildimage/pull/4388

**Pull request merged**